File: prlsp/git.py
"""Git repo/branch/remote detection."""

import re
import subprocess
from dataclasses import dataclass


@dataclass
class GitInfo:
    root: str
    branch: str
    owner: str
    repo: str


def _run_git(args: list[str], cwd: str) -> str | None:
    try:
        result = subprocess.run(
            ["git"] + args,
            cwd=cwd,
            capture_output=True,
            text=True,
            timeout=5,
        )
        if result.returncode == 0:
            return result.stdout.strip()
    except (subprocess.TimeoutExpired, FileNotFoundError):
        pass
    return None


def _parse_remote(url: str) -> tuple[str, str] | None:
    # SSH: git@github.com:owner/repo.git
    m = re.match(r"git@github\.com:([^/]+)/([^/]+?)(?:\.git)?$", url)
    if m:
        return m.group(1), m.group(2)
    # HTTPS: https://github.com/owner/repo.git
    m = re.match(r"https://github\.com/([^/]+)/([^/]+?)(?:\.git)?$", url)
    if m:
        return m.group(1), m.group(2)
    return None


def detect_git_info(workspace_path: str) -> GitInfo | None:
    root = _run_git(["rev-parse", "--show-toplevel"], workspace_path)
    if not root:
        return None

    branch = _run_git(["rev-parse", "--abbrev-ref", "HEAD"], workspace_path)
    if not branch:
        return None

    remote_url = _run_git(["remote", "get-url", "origin"], workspace_path)
    if not remote_url:
        return None

    parsed = _parse_remote(remote_url)
    if not parsed:
        return None

    owner, repo = parsed
    return GitInfo(root=root, branch=branch, owner=owner, repo=repo)

File: go/git.go
package main

import (
	"os/exec"
	"regexp"
	"strings"
)

type GitInfo struct {
	Root   string
	Branch string
	Owner  string
	Repo   string
}

func runGit(args []string, cwd string) string {
	cmd := exec.Command("git", args...)
	cmd.Dir = cwd
	out, err := cmd.Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

var (
	sshRemoteRe   = regexp.MustCompile(`^git@github\.com:([^/]+)/([^/]+?)(?:\.git)?$`)
	httpsRemoteRe = regexp.MustCompile(`^https://(?:[^.@/]+@)?github\.com/([^/]+)/([^/]+?)(?:\.git)?$`)
)

func parseRemote(url string) (owner, repo string, ok bool) {
	if m := sshRemoteRe.FindStringSubmatch(url); m != nil {
		return m[1], m[2], true
	}
	if m := httpsRemoteRe.FindStringSubmatch(url); m != nil {
		return m[1], m[2], true
	}
	return "", "", false
}

func detectGitInfo(workspacePath string) *GitInfo {
	root := runGit([]string{"rev-parse", "--show-toplevel"}, workspacePath)
	if root == "" {
		return nil
	}
	branch := runGit([]string{"rev-parse", "--abbrev-ref", "HEAD"}, workspacePath)
	if branch == "" {
		return nil
	}
	remoteURL := runGit([]string{"remote", "get-url", "origin"}, workspacePath)
	if remoteURL == "" {
		return nil
	}
	owner, repo, ok := parseRemote(remoteURL)
	if !ok {
		return nil
	}
	return &GitInfo{Root: root, Branch: branch, Owner: owner, Repo: repo}
}
